### PR TITLE
Decimal expansion calculator

### PR DESCRIPTION
**Pull request by umvegas**
# Goodie Pull Request Template

Using this template will help us better understand your Instant Answer and assist you when necessary.  Simply copy and paste the markdown below into the description of your GitHub pull request and complete as appropriate.

**What does your Instant Answer do?**
Computes the complete decimal expansion for a quotient of two whole numbers, including an indication of which digits are repeated if the answer is a repeating decimal.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
Neither organic links nor other instant answer calculators give a full decimal expansion for a quotient, nor do they give information about repetition in the answer.

**What is the data source for your Instant Answer? (Provide a link if possible)**
No source. It's a code-only answer.

**Why did you choose this data source?**
N/A

**Are there any other alternative (better) data sources?**
There are of course, other calculators, including other instant answers. See the above answer for why I want to add another.

**What are some example queries that trigger this Instant Answer?**
divide 22 by 7
longdiv 22 / 7
decimate 22 by 7

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
This instant answer is mostly a novelty, but doesn't seem to have been produced already.

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No, not that I know of.

**Which existing Instant Answers will this one supercede/overlap with?**
It won't supersede any that I know of. It could probably be an extension of the current calculator instant answer, but it's simpler to have it be a separate project.

**Are you having any problems? Do you need our help with anything?**
I think I've been able to follow the guidelines, but I guess your review will have to decide that.

**Where did you hear about DuckDuckHack? (For first time contributors)**
From the main search page.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
decimate_screenshot

## Checklist
Please place an 'X' where appropriate.

```
- [X] Added metadata and attribution information (https://duck.co/duckduckhack/metadata)
- [X] Wrote test file and added to t/ directory (https://duck.co/duckduckhack/goodie_tests)
- [X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
- [X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
- [X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        - [X] Google Chrome
        - [X] Firefox
        - [] Opera
        - [X] IE 10

    - Windows 7
        - [X] Google Chrome
        - [X] Firefox
        - [] Opera
        - [] IE 8
        - [] IE 9
        - [X] IE 10

    - Windows XP
        - [] IE 7
        - [] IE 8

    - Mac OSX
        - [] Google Chrome
        - [] Firefox
        - [] Opera
        - [] Safari

    - iOS (iPhone)
        - [X] Safari Mobile
        - [] Google Chrome
        - [] Opera

    - iOS (iPad)
        - [X] Safari Mobile
        - [X] Google Chrome
        - [] Opera

    - Android
        - [] Firefox
        - [] Native Browser
        - [] Google Chrome
        - [] Opera

```

### CODE REVIEW

File: lib/DDG/Goodie/Decimate.pm [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+use DDG::Goodie;
+use strict;
+
+zci answer_type => "decimate";
+zci is_cached   => 1;
+
+# Metadata.  See https://duck.co/duckduckhack/metadata for help in filling out this section.
+name "Decimate";
+description "Calculates complete decimal expansion of a quotient, showing the repetend.";
+primary_example_queries "decimate 2473 / 23";
```
Comment (mintsoft) [inline lib/DDG/Goodie/Decimate.pm]: I'm not sure this name is accurate, decimate means reduce by 1/10th

File: lib/DDG/Goodie/Decimate.pm [hunk 2]
```diff
@@ -0,0 +1,99 @@
+package DDG::Goodie::Decimate;
+# ABSTRACT: Write an abstract here
```
Comment (mintsoft) [inline lib/DDG/Goodie/Decimate.pm]: Can you move the abstract from the line below onto this one, so it's:
```
# ABSTRACT: Do a long ...
```

File: lib/DDG/Goodie/Decimate.pm [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+# Do a long division, showing a complete decimal expansion, including the repetend.
+
+use DDG::Goodie;
+use strict;
+
+zci answer_type => "decimate";
+zci is_cached   => 1;
+
+# Metadata.  See https://duck.co/duckduckhack/metadata for help in filling out this section.
```
Comment (mintsoft) [inline lib/DDG/Goodie/Decimate.pm]: you can bomb any comment that simply explains what each section is for

File: lib/DDG/Goodie/Decimate.pm [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+            twitter => "\@umvegas";
+
+# Triggers
+triggers start => "decimate", "divide", "longdiv", "long div", "long divide", "long division";
+
+# Handle statement
+handle remainder => sub {
+
+    # optional - regex guard
+    return unless qr/^\w+/;
```
Comment (mintsoft) [inline lib/DDG/Goodie/Decimate.pm]: I think this is probably unnecessary as https://github.com/duckduckgo/zeroclickinfo-goodies/pull/1164/files#diff-9cae36265bcb8bff58ef33306d50feb8R35 will guard the query

File: lib/DDG/Goodie/Decimate.pm [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    
+    return unless $_; # Guard against "no answer"
+    
+    my @ml = /(\d+)\s*(?:\/|by)\s*(\d+)/; # numerator / denominator, "/" or "by"
+    return unless @ml;                    # quit if no match
+    
+    my $numer = $ml[0];         # the numerator
```
Comment (mintsoft) [inline lib/DDG/Goodie/Decimate.pm]: can you rename this variable to `$numerator` and the one below as `$denominator`, that way these comments can be removed

File: lib/DDG/Goodie/Decimate.pm [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    my $denom = $ml[1];         # the denominator
+    my @nl = split("", $numer); # list of digits in the numerator
+    my $dotAt = @nl;            # position of the decimal in the numerator
```
Comment (mintsoft) [inline lib/DDG/Goodie/Decimate.pm]: I think most of these variables should be renamed to explain their purpose better, probably based on the content in the comments

File: lib/DDG/Goodie/Decimate.pm (hunk 1 above)
Comment (umvegas) [inline lib/DDG/Goodie/Decimate.pm]: It was intended as a play on words, but if it doesn't work I don't mind changing it. I think the original usage was re: Roman armies, so I'm not too concerned with inaccurate usage.

File: lib/DDG/Goodie/Decimate.pm (hunk 6 above)
Comment (umvegas) [inline lib/DDG/Goodie/Decimate.pm]: Thanks for the comments. I'll rework things. How do I let you know that I've made changes and I want you to re-evaluate?

**Comment by moollaza** [comment]
@umvegas thanks a lot for taking the time to contribute, we really appreciate it!

I really wish you would have [contacted](https://duck.co/duckduckhack/ddh-intro#submitting-your-idea) us for approval though, because we're *just* about to receive a Pull Request for a new Calculator Instant Answer, which will be able to handle division (and most other math).

However, I don't believe it will initially handle queries with words (i.e. "divide X by Y", rather "calculate X/Y") so if you're interested in improving it once it's gone live we'd certainly welcome any contributions!

We *may* be able to launch this depending on how long the Calculator will take to go live (hopefully it will be live very soon!), but please know that ultimately this code will be replaced. Our goal is to unify all the existing Math related IAs and ideally offload all the calculation to an API (or Math lib) to simplify things.

How does that sound?

Thanks again for taking the time to contribute, we really do appreciate it and hope you'll stick around!

**Comment by umvegas** [comment]
Sounds good. Thanks.

**Pull request closed without merging**